forc-pkg/src/pkg.rs

```diff
@@ -2556,22 +2556,30 @@ pub fn build(
 
 /// Standardize the JSON ABI data structure by eliminating duplicate types. This is an iterative
 /// process because every time two types are merged, new opportunities for more merging arise.
-fn standardize_json_abi_types(json_abi_program: &mut JsonABIProgram) {
+fn standardize_json_abi_types(json_abi_program: &mut fuels_types::ProgramABI) {
     loop {
         // If type with id_1 is a duplicate of type with id_2, then keep track of the mapping
         // between id_1 and id_2 in the HashMap below.
         let mut old_to_new_id: HashMap<usize, usize> = HashMap::new();
 
-        // HashSet to eliminate duplicate type declarations.
-        let mut types_set: HashSet<JsonTypeDeclaration> = HashSet::new();
-
-        // Insert values in the HashSet `types_set` if they haven't been inserted before.
-        // Otherwise, create an appropriate mapping in the HashMap `old_to_new_id`.
-        for decl in json_abi_program.types.iter_mut() {
-            if let Some(ty) = types_set.get(decl) {
+        // A vector containing unique `fuels_types::TypeDeclaration`s.
+        //
+        // Two `fuels_types::TypeDeclaration` are deemed the same if the have the same
+        // `type_field`, `components`, and `type_parameters` (even if their `type_id`s are
+        // different).
+        let mut deduped_types: Vec<fuels_types::TypeDeclaration> = Vec::new();
+
+        // Insert values in `deduped_types` if they haven't been inserted before. Otherwise, create
+        // an appropriate mapping between type IDs in the HashMap `old_to_new_id`.
+        for decl in json_abi_program.types.iter() {
+            if let Some(ty) = deduped_types.iter().find(|d| {
+                d.type_field == decl.type_field
+                    && d.components == decl.components
+                    && d.type_parameters == decl.type_parameters
+            }) {
```

Comment:
Does `fuels_types::TypeDeclaration` need `Hash`?  Is it too fiddly to introduce?

Reply:
Yeah we'd need to have custom implementations of `Hash` and `PartialEq` in `fuels-types` (like I had in `sway-types`).